### PR TITLE
[WIP] WebSocket updates

### PR DESCRIPTION
**Pull request by pavel-kokolemin**
- Add session resuming to WsDataConnection/WsServerConnection.
  Needed when Cloudflare servers restarted.
  Reimplement something that we had with ZMQ (transparent TCP restarts)
- Add Bip*Connection that could handle BIP encryption for other connections.
  Superfluous TransportBIP15x* could be removed later.
- Fix BIP15X rekeying regression from client side.
- Remove adding Transport to ZMQ streams.

### CODE REVIEW

**Comment by sergey-chernikov** [comment]
Some general notes on the PR family:
1. BIP is a Bitcoin Improvement Proposal, IMO better to precise file and class naming, 15x suffix seems appropriate
2. You use generic transport and connection in e.g. BipConnection (it can be ZMQ + SSL at some point) - why not name it something like CombinedTransportConnection or just TransportConnection?
3. We need working tests (TestNetwork in the terminal repo) to be sure all edge cases are covered properly.

**Comment by pavel-kokolemin** [comment]
>     1. BIP is a Bitcoin Improvement Proposal, IMO better to precise file and class naming, 15x suffix seems appropriate
Sure, fixed. I just don't like that it's too verbose.
>     2. You use generic transport and connection in e.g. BipConnection (it can be ZMQ + SSL at some point) - why not name it something like CombinedTransportConnection or just TransportConnection?
My plan is to remove TransportConnection (merge it with `Bip15x*Connection`) someday, so there will be only `Bip15x*Connection` that will only work above `Ws*Connection`. SSL will be handled by `WsDataConnection` only. We won't be using ZMQ in new code. So I guess we could leave it as is.

>     3. We need working tests (TestNetwork in the terminal repo) to be sure all edge cases are covered properly.
I'm working on that. I don't expect any problems here, so I guess this could be merged to unblock others (as staging is broken now).

**Comment by pavel-kokolemin** [comment]
Network unit tests reverted, except:
- `ZMQ_BIP15X_MalformedData` is not relevant now and removed
- `ZMQ_BIP15X_ConnectionTimeout` will be fixed a bit later

**Comment by sergey-chernikov** [comment]
2. The idea behind connection and transport separation was to encapsulate all connection instances into base type and don't care about BIP15x specifics when just sending/receiving data. All logic was put to transport. If you gonna merge it back, you'll have to declare BIP15x-specific connection types again everywhere

File: BlocksettleNetworkingLib/WsDataConnection.cpp [hunk 1]
```diff
@@ -135,9 +150,9 @@ int WsDataConnection::callback(lws *wsi, int reason, void *user, void *in, size_
 {
    switch (reason) {
       case LWS_CALLBACK_OPENSSL_LOAD_EXTRA_CLIENT_VERIFY_CERTS: {
-//         if (!params_.useSsl || params_.caBundleSize == 0) {
+         if (!params_.useSsl || params_.caBundleSize == 0) {
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/WsDataConnection.cpp:153]: Why do we need SSL re-enabled? It should be superseded by BIP15x

File: BlocksettleNetworkingLib/WsServerConnection.cpp [hunk 2]
```diff
@@ -108,7 +108,11 @@ void WsServerConnection::stopServer()
       return;
    }
 
-   stopped_ = true;
+   DataToSend toSend{kAllClientsId, WsPacket::close()};
+   {  std::lock_guard<std::mutex> lock(mutex_);
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/WsServerConnection.cpp:112]: Please put `{` in block and code line to separate rows

File: BlocksettleNetworkingLib/WsServerConnection.cpp [hunk 3]
```diff
@@ -118,146 +122,378 @@ void WsServerConnection::stopServer()
...
-            return -1;
-         }
-         if (rc != static_cast<int>(packet.getSize())) {
-             SPDLOG_LOGGER_ERROR(logger_, "write truncated");
-             return -1;
-         }
-
-         if (!client.packets.empty()) {
-            lws_callback_on_writable(client.wsi);
+         auto &connection = connections_.at(wsi);
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/WsServerConnection.cpp:314]: at() may throw under some conditions - do you catch elsewhere?

File: BlocksettleNetworkingLib/WsServerConnection.cpp [hunk 4]
```diff
@@ -118,146 +122,378 @@ void WsServerConnection::stopServer()
...
 }
 
-std::thread::id WsServerConnection::listenThreadId() const
+bool WsServerConnection::done() const
 {
-   return listenThread_.get_id();
+   if (!shuttingDown_) {
+      return false;
+   }
+   {  std::lock_guard<std::mutex> lock(mutex_);
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/WsServerConnection.cpp:420]: Same here for `{` and new line

File: BlocksettleNetworkingLib/WsServerConnection.cpp [hunk 5]
```diff
@@ -118,146 +122,378 @@ void WsServerConnection::stopServer()
...
-   // So we use multicast here.
-   size_t cntClients = clients_.size();
-   for (const auto &item : clients_) {
-      if (SendDataToClient(item.first, data)) {
-         cntClients--;
-      }
-   }
-   return (cntClients == 0);
+{
+   return SendDataToClient(kAllClientsId, data);
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/WsServerConnection.cpp:496]: Did you read the deleted comment? Did you test the broadcast of encrypted data?

File: BlocksettleNetworkingLib/ZmqServerConnection.cpp [hunk 6]
```diff
@@ -452,30 +367,26 @@ bool ZmqServerConnection::QueueDataToSend(const std::string& clientId, const std
...
 {
-   if (transport_) {
-      transport_->periodicCheck();
-   }
 }
 
 void ZmqServerConnection::SendDataToDataSocket()
 {
-   decltype(dataQueue_) pendingData;
+   std::deque<DataToSend> pendingData;
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/ZmqServerConnection.cpp:374]: Why this change? From my perspective the deleted line is more obvious and clear, that it's a destination for swap of `dataQueue_`.

File: BlocksettleNetworkingLib/ZmqServerConnection.cpp [hunk 7]
```diff
@@ -452,30 +367,26 @@ bool ZmqServerConnection::QueueDataToSend(const std::string& clientId, const std
...
+
    {
       FastLock locker{dataQueueLock_};
       pendingData.swap(dataQueue_);
    }
 
    for (const auto &dataPacket : pendingData) {
-      int result = zmq_send(dataSocket_.get(), dataPacket.clientId.c_str()
-         , dataPacket.clientId.size(), ZMQ_SNDMORE);
+      int result = zmq_send(dataSocket_.get(), dataPacket.clientId.c_str(), dataPacket.clientId.size(), ZMQ_SNDMORE);
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/ZmqServerConnection.cpp:382]: What this and next change actually does? Except making this line longer and scrollable on small displays.

File: BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp [hunk 8]
```diff
@@ -9,18 +9,17 @@
...
-#include <zmq.h>
-#include <spdlog/spdlog.h>
+
 #include "ActiveStreamClient.h"
 #include "FastLock.h"
 #include "MessageHolder.h"
-#include "Transport.h"
 
+#include <zmq.h>
+#include <spdlog/spdlog.h>
```
Comment (sergey-chernikov) [inline BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp:18]: Again - is this a blind rollback of everything? Regardless of useful changes?

File: BlocksettleNetworkingLib/HeadlessContainer.cpp [hunk 9]
```diff
@@ -1338,13 +1339,6 @@ bool RemoteSigner::Start()
       return true;
    }
 
-   if (opMode() == OpMode::RemoteInproc) {
```
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/HeadlessContainer.cpp:1341]: Everything works fine and `RemoteInproc` was not used, `LocalInproc` should be enough.
> I mean, any kind of inproc connection (even not a localhost one)?
Sorry, I don't understand the question. Why would you need `RemoteInproc` type? If you want in-process signer then local connection is the sane type to use.

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp [hunk 10]
```diff
@@ -216,112 +211,69 @@ std::unique_ptr<BIP15xPeer> TransportBIP15xServer::getClientKey(const std::strin
...
+   auto payload = BinaryData::fromString(encData);
 
-      // If we're still handshaking, take the next step. (No fragments allowed.)
-      if (msg.getType() == bip15x::MsgType::Heartbeat) {
-         UpdateClientHeartbeatTimestamp(clientID);
+   // Decrypt only if the BIP 151 handshake is complete.
+   if (connData->bip151HandshakeCompleted_) {
+      //decrypt packet
+      auto result = connData->encData_->decryptPacket(
+         payload.getPtr(), payload.getSize(),
```
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:231]: `Ws*Connection` does not use fragmented packets.

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp [hunk 11]
```diff
@@ -216,112 +211,69 @@ std::unique_ptr<BIP15xPeer> TransportBIP15xServer::getClientKey(const std::strin
...
-            logger_->error("[TransportBIP15xServer::processIncomingData] encryption"
-               " handshake is incomplete");
-         }
-         if (clientErrorCb_) {
-            clientErrorCb_(clientID, "encryption handshake failure");
-         }
+   if (msg.getType() > bip15x::MsgType::AEAD_Threshold) {
+      if (!processAEADHandshake(msg, clientID)) {
+         reportFatalError(connData);
          return;
```
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:258]: I don't know, it was already there, just checked old `ZMQ_BIP15X_ServerConnection.cpp`.

File: BlocksettleNetworkingLib/WsDataConnection.cpp (hunk 1 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:153]: `WsDataConnection` could be used without BIP15x.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 2 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp]: Sure, updated.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 3 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:316]: `at` here is on purpose for sanity check

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 4 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp]: Sure, fixed.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: `WsServerConnection` does not broadcast encrypted data now, that is up to `Bip15xServerConnection` and it will encrypt it as needed.

File: BlocksettleNetworkingLib/ZmqServerConnection.cpp (hunk 6 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/ZmqServerConnection.cpp:374]: Sure, reverted.

File: BlocksettleNetworkingLib/ZmqServerConnection.cpp (hunk 7 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/ZmqServerConnection.cpp:382]: Just reverted old file.

File: BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp (hunk 8 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp:18]: What exactly we are missing here? ZMQ will be deprecated and removed soon.

**Comment by pavel-kokolemin** [comment]
> The idea behind connection and transport separation was to encapsulate all connection instances into base type and don't care about BIP15x specifics when just sending/receiving data.

Yes, I got that and still don't see the point.

> you'll have to declare BIP15x-specific connection types again everywhere

Sorry, I don't see why I would need to do that.

File: BlocksettleNetworkingLib/HeadlessContainer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/HeadlessContainer.cpp:1341]: It's usable, but previously it was in-memory connection from ZMQ - is it possible to use it now?

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:228]: Did you notice ZMQ in my comment? I'm talking generally, not only WS

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:255]: Ofc, but you added new return below - better to remove one, that is unnecessary.

File: BlocksettleNetworkingLib/WsDataConnection.cpp (hunk 1 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:153]: It could, but should it?

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 3 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:316]: `at` can throw an exception - is it caught somewhere? Or is it ok to crash the whole process?

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: So, you're saying that Bip15xServerConnection can *broadcast* data to multiple clients, encrypted once? Encrypted by what client's key?

File: BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp (hunk 8 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/ZmqStreamServerConnection.cpp:18]: Just the changes that are not necessary to revert. Headers reorder in this case.

**Comment by sergey-chernikov** [comment]
> Yes, I got that and still don't see the point.

OOD, encapsulation. Other classes need to know about the connection as less as possible. If you declare a BIP15x-specific connection and start to use its specific methods, no other replacement for it will be possible to use without changing all classes that use it.

File: BlocksettleNetworkingLib/WsDataConnection.cpp (hunk 1 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:153]: I think yes, where we don't need to bother with all that Bip15x nonsense. Whole world uses SSL while we trying to reinvent wheel.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: Bip15xServerConnection will send encrypted packets to separate clients. And this is tested in unit tests.

**Comment by pavel-kokolemin** [comment]
> > Yes, I got that and still don't see the point.
> 
> OOD, encapsulation. Other classes need to know about the connection as less as possible. If you declare a BIP15x-specific connection and start to use its specific methods, no other replacement for it will be possible to use without changing all classes that use it.

We already have `DataConnection` and `ServerConnection` as abstraction. And so nobody uses BIP15x-specific stuff unless strictly needed.

File: BlocksettleNetworkingLib/WsDataConnection.cpp (hunk 1 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:153]: Strong statement regarding BIP15x nonsense...
@goatpig please decide here

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 3 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:316]: Better to crash whole process in case of logic errors than to have security breach as result of UB exploitation. So yes, much better.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: Then this shouldn't be called *broadcast*. What's the real broadcast for in WsServerConnection then?

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:255]: Yes, sure, fixed.

**Comment by sergey-chernikov** [comment]
> We already have `DataConnection` and `ServerConnection` as abstraction. And so nobody uses BIP15x-specific stuff unless strictly needed.

Yes, this is what I was trying to achieve. But in case you obsolete this state of things, you will start using specific BIP15x methods (e.g. to manage keys) in a connection instead of the transport.

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:228]: Yes, I saw it, but we will not ZMQ with BIP15X.

[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:228]: What if it will be just plain TCP socket? Should we re-implement the transport?

File: BlocksettleNetworkingLib/HeadlessContainer.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/HeadlessContainer.cpp:1341]: No, we don't have in-memory connection right now. I guess using loopback address should work here.

**Comment by pavel-kokolemin** [comment]
> you will start using specific BIP15x methods (e.g. to manage keys) in a connection instead of the transport.

Yes, sure, but I don't see the problem. If you need to manage keys you will need to do it somewhere. Right now it is encapsulated in `Bip15x*Connection`. If you don't need to manage keys you use abstract `DataConnection`/`ServerConnection`. If we ever need to you to use BIP15x over ZMQ again, we will use `Bip15x*Connection` with `Zmq*Connection` and `Zmq*Connection` will know nothing about BIP15x. And there is won't be need to add bunch of methods and new classes here and there like we just had.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 3 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:316]: I wasn't asking if it's better - only if it's ok or not. Not sure though if we can't do it more gracefully - but ok, let's keep it as is.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: `WsServerConnection` does not know and should not know about any of that, if it got command to send packets to all clients it will do so. `Bip15xServerConnection` should do what is needed here (encrypt packets separately).

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:255]: Not yet, I guess

File: BlocksettleNetworkingLib/TransportBIP15xServer.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/TransportBIP15xServer.cpp:228]: Right, perhaps in such case using Transport would be a good idea to hide some of that `BIP15x` code. But I really doubt that is a good idea. We won't be able use Cloudflare for DDoS protection, nginx for routing etc.

File: BlocksettleNetworkingLib/WsDataConnection.cpp (hunk 1 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:153]: We spend too much time here trying to solve already solved problem. And implementing home brew encryption is very rarely a good idea. For bitcoin BIP15X might have sense as it's decentralized. We don't need any of that.

Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsDataConnection.cpp:154]: For example we have had a bug when memory was beyond heap in `BIP151Session::decPayload`. That was accidental detected by ASAN. It's direct way for security exploitation. That bug was fixed but who knows what else we had here.

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (sergey-chernikov) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:500]: The question here - will it do so properly? You have a special receiver to maintain broadcast - is it compatible with SendDataToAllClients in Bip15ServerConnection? And if it's not used - what's the use case for kAllClientsId at all?

File: BlocksettleNetworkingLib/ZmqServerConnection.cpp [hunk 12]
```diff
@@ -230,24 +164,21 @@ void ZmqServerConnection::listenFunction()
...
-   poll_items[ZmqServerConnection::MonitorSocketIndex].events = ZMQ_POLLIN;
-
    logger_->debug("[{}] poll thread started for {}", __func__, connectionName_);
 
    int errorCount = 0;
 
    while (true) {
       int periodMs = std::chrono::duration_cast<std::chrono::milliseconds>(kHearthbeatCheckPeriod).count();
-      int result = zmq_poll(poll_items, 3, periodMs);
+      int result = zmq_poll(poll_items, sizeof(poll_items) / sizeof(poll_items[0]), periodMs);
```
Comment (Ation) [inline BlocksettleNetworkingLib/ZmqServerConnection.cpp:181]: You have compile time variable. Make it constexpr, don't use 30 years old C tricks.

Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/ZmqServerConnection.cpp:181]: Sorry, I don't understand what are you referring to. How `constexpr` will help here?

File: BlocksettleNetworkingLib/WsServerConnection.cpp (hunk 5 above)
Comment (pavel-kokolemin) [reply BlocksettleNetworkingLib/WsServerConnection.cpp:518]: > The question here - will it do so properly?
It will do so properly as far as it concerns `WsServerConnection`. Not sure what are we discussing here, unit tests cover this.

**Comment by pavel-kokolemin** [comment]
PR redirected to `auth_address_tiers` branch - https://github.com/BlockSettle/common/pull/2022
